qualtran/_infra/controlled.py

```diff
@@ -318,6 +380,43 @@ def on_classical_vals(self, **vals: 'ClassicalValT') -> Dict[str, 'ClassicalValT
 
         return vals
 
+    def add_my_tensors(
+        self,
+        tn: 'qtn.TensorNetwork',
+        tag: Any,
+        *,
+        incoming: Dict[str, 'SoquetT'],
+        outgoing: Dict[str, 'SoquetT'],
+    ):
+        from qualtran._infra.composite_bloq import _flatten_soquet_collection
+        from qualtran.cirq_interop._cirq_to_bloq import (
+            _tensor_data_from_unitary_and_signature,
+            _tensor_shape_from_signature,
+        )
+
+        # Create an identity tensor corresponding to the signature of current Bloq
+        data = _tensor_data_from_unitary_and_signature(
+            np.eye(2 ** self.signature.n_qubits(), dtype=np.complex128), self.signature
+        )
+        # Verify it has the right shape
+        in_ind = _flatten_soquet_collection(incoming[reg.name] for reg in self.signature.lefts())
+        out_ind = _flatten_soquet_collection(outgoing[reg.name] for reg in self.signature.rights())
+        assert data.shape == tuple(2**soq.reg.bitsize for ind in [out_ind, in_ind] for soq in ind)
+        # Figure out the ctrl indexes for which the ctrl is "active"
+        active_idx: List[Union[int, slice]] = [slice(x) for x in data.shape]
+        ctrl_idx = 0
+        for cv, ctrl_reg in zip(self.ctrl_spec.cvs, self.ctrl_regs):
```

Comment:
if I understand correctly, this is highly specific to the current concrete implementation of `CtrlSpec`. You can imagine a world where there is an abstract `CtrlSpec` base class and multiple implementations that can support things other than conjunction of a bunch of equals-a-value. In my head, `cvs` would be an implementation detail and `CtrlSpec.is_active` would be the interface.

Reply:
At that point, we could iterate over all `2 ** n` possible bitstrings that `cvs` can take and call `is_active` to check if that subspace should be "Activated"

Reply:
Note `SumOfProducts` is a generalized representation that can be used to represent any arbitrary boolean function.